atomic-factory/src/main/java/org/infinispan/atomic/AtomicObjectFactory.java

```diff
@@ -0,0 +1,264 @@
+import org.infinispan.atomic.container.Container;
+import org.infinispan.atomic.container.ContainerSignature;
+import org.infinispan.util.logging.Log;
+import org.infinispan.util.logging.LogFactory;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.lang.reflect.Method;
+import java.util.*;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+
+/**
+ * @author Pierre Sutra
+ * @since 7.0
+ */
+public class AtomicObjectFactory {
+
+   //
+   // CLASS FIELDS
+   //
+   private static Log log = LogFactory.getLog(AtomicObjectFactory.class);
+   private static Map<Cache,AtomicObjectFactory> factories = new HashMap<>();
+   public synchronized static AtomicObjectFactory forCache(Cache cache){
+      if(!factories.containsKey(cache))
+         factories.put(cache, new AtomicObjectFactory(cache));
+      return factories.get(cache);
+   }
+   protected static final int MAX_CONTAINERS=1000;// 0 means no limit
+   public static final Map<Class,List<String>> updateMethods;
+   static{
+      updateMethods = new HashMap<>();
+
+      updateMethods.put(List.class, new ArrayList<String>());
+      updateMethods.get(List.class).add("add");
+      updateMethods.get(List.class).add("addAll");
+
+      updateMethods.put(Set.class, new ArrayList<String>());
+      updateMethods.get(Set.class).add("add");
+      updateMethods.get(Set.class).add("addAll");
+
+      updateMethods.put(Map.class, new ArrayList<String>());
+      updateMethods.get(Map.class).add("put");
+      updateMethods.get(Map.class).add("putAll");
+   };
+
+
+
+   //
+   // OBJECT FIELDS
+   //
+   private Cache cache;
+   private Map<ContainerSignature,Container> registeredContainers;
+   private int maxSize;
+   private final ExecutorService evictionExecutor = Executors.newCachedThreadPool();
+
+
+   /**
+    *
+    * Returns an object factory built on top of cache <i>c</i> with a bounded amount <i>m</i> of
+    * containers in it. Upon the removal of a container, the object is stored persistently in the cache.
+    *
+    * @param c it must be synchronous.and transactional (with autoCommit set to true, its default value).
+    * @param m max amount of containers kept by this factory.
+    * @throws InvalidCacheUsageException
+    */
+   public AtomicObjectFactory(Cache<Object, Object> c, int m) throws InvalidCacheUsageException{
+      cache = c;
+      maxSize = m;
+      registeredContainers= new LinkedHashMap<ContainerSignature,Container>(){
+         @Override
+         protected boolean removeEldestEntry(final java.util.Map.Entry<ContainerSignature,Container> eldest) {
+            if(maxSize!=0 && this.size() >= maxSize){
+               evictionExecutor.submit(new Callable<Void>() {
+                  @Override
+                  public Void call() throws IOException {
+                     try {
+                        log.debug("Disposing " + eldest.getValue().toString());
```

Comment:
wrap the call with a log.isDebugEnabled please, and use the Formatter-enabled methods